### ISSUE TITLE
Index out of range should include the index

### ISSUE THREAD

**Issue by terrajobst** [dotnet/corefx#4940]
When indexing into collections with an invalid index, we throw an `IndexOfRangeException`. In the spirit of being helpful, it would be tremendously useful if the exception message would include the actual index. This might not be easily possible in all cases, for instance, if the CLR throws it when indexing into an array but it seems borderline malicious to generally not include the index.

Also, we've recently fixed this behavior for duplicate key exceptions, so adding this seems to be consistent.

**Comment by AlgorithmsAreCool** [comment]
Is this for System.Collections.* namespace? Or some subset or superset of it?

**Comment by Foda** [comment]
I agree that this would be a good addition. Seems like a change similar to https://github.com/dotnet/coreclr/pull/1452 would be appropriate. I created a [branch](https://github.com/Foda/corefx/commit/c65c929080a58270b734db5f3174fa4ee8ca382d) with some simple changes that we can expand on and discuss if it should be the path forward.

**Comment by JonHanna** [comment]
It could though cause some passed-through exceptions to report the wrong index, if perhaps offsets were involved.

**Comment by JonHanna** [comment]
A fuller example of my point. Consider:

```C#
public T this[int index]
{
    get
    {
        int len = _first.Count;
        return index < len ? _first[index] : _second[index - len];
    }
}
```

Right now that would be correct code that either works correctly or gives a reasonable exception. If `_second` is a type that includes indices in out of range exceptions then e.g if `_first` and `_second` both have 10 items, making the combined list 20 in size and code looks for the item with index `23` results in an `ArgumentOutOfRange` that says the `"index"` parameter was at fault.

If `_second[]` included the index passed in its exception, then this becomes an `ArgumentOutOfRange` that says the `"index"` was at fault because it was `3`, despite 3 being well within the range [0, 19] and so cause more confusion than including an index could resolve.

**Comment by hughbe** [comment]
This is a reasonable criticism of the proses change, but perhaps It could be argued that it is up to the API writer to provide custom argument checking anyway for this kind of function?

**Comment by JonHanna** [comment]
Who's to say they didn't? They could have either spent considerable time verifying that every possible incorrect input threw an appropriate exception, or indeed started with an explicit guard and then later verified that without it every possible incorrect input threw an appropriate exception, and hence removed that explicit check as an optimisation.

**Comment by hughbe** [comment]
Sorry if i wasn't clear - basically this check improves the vast majority of areas when this exception is thrown. 
In the edge cases you described, it may be suggestible for the writer of the API to add her own argument validation with a different ArgumentOutOfRangeException objectValue and message.

**Comment by JonHanna** [comment]
Yes, but that's true of all backwards-compatibility issues. They'd all go away if people changed the code that worked before they were introduced.

**Comment by hughbe** [comment]
Oh good point I hadn't considered backwards compatibility issues!
Up to you guys to decide - I assume discussion ha already been done as this was marked up for grabs

**Comment by JonHanna** [comment]
I'm not at all clear on that myself. I do think the balance of usefulness is such that I'd rather it remained as is. (In this regard, you spotted a few cases where the error message was incorrect, and I spotted one more looking at your PR, I think fixing those would be something it would be hard to see as anything other than completely positive).

**Comment by svick** [comment]
@JonHanna Maybe modifying the exception message could help with that confusing case? Specifically, I was thinking about including the type of the collection there. So, in your example, the message would be something like:

> Index was out of range when accessing System.Int32[], the invalid value was 3.

But since you're accessing `SomeCustomCollection<int>`, not `int[]`, this might help you realize that the indexes are different.

**Comment by terrajobst** [comment]
@JonHanna: The point is well taken but I think in most cases the stack trace will disambiguate. Your indexer is clearly a custom type. If this type isn't at the top of the exceptions's stack the index out of range doens't apply to it.

In other words, this problem of locality is true for all exception types, for instance, argument exceptions. I don't see why this should be an argument against exposing the index.

@JonHanna: I don't see any compatibility issues. We'd introduce a new ctor that would take an index. Existing code will continue to throw an exception whose message is the same as today. New code is able to call the overload that provides the index. What am i missing?

**Comment by JonHanna** [comment]
> Maybe modifying the exception message could help with that confusing case?

I think that would be worse.

I've been musing on this. My objection is that it makes an existing case that behaves perfectly, behave imperfectly. However, on balance:

1. The case is relatively rare.
2. Nothing is fully "broken", just imperfect.
3. The impact of that imperfection would (almost?) only be on debugging situations only.
4. That impact will perhaps cause some brief confusion, but the confusion should be quickly resolved.
5. Such debugging situations are exactly the same as would likely benefit from this change the vast majority of the time.

As such, I think the objection I made is pretty minimal in impact, and this idea a good one.

**Comment by hughbe** [comment]
This can be closed now. Fixed by PR #5672

**Comment by terrajobst** [comment]
Agreed.
